sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -203,3 +203,28 @@ abstract class BinaryNode extends LogicalPlan {
 abstract class OrderPreservingUnaryNode extends UnaryNode {
   override final def outputOrdering: Seq[SortOrder] = child.outputOrdering
 }
+
+object LogicalPlanIntegrity {
```

Comment:
shall we check the physical plan as well?

Reply:
Currently, `RuleExecutor` is not used for transforming physical plans now: https://github.com/apache/spark/blob/a1e459ed9f6777fb8d5a2d09fda666402f9230b9/sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala#L363-L367

Is it okay to replace the code above with `RuleExecutor` in this PR?

Reply:
I think we in general won't change output in physical plan?

Reply:
Ah, I see. Nice suggestion. Looked around the related code, I think it is not easy to catch all the ill-formed case as you said, but IMO most common patterns to cause duplicate `ExprId`s are a reuse of the `ExprId`s of references, e.g., `a#1 + 1 as a#1` as you pointed above. So, how about just checking this reuse pattern in the plan integrity check?  https://github.com/apache/spark/pull/29585/files#diff-27c76f96a7b2733ecfd6f46a1716e153R238